wpengine/atlas-content-modeler
# [MTKA-1389] Export ACM Blueprints via WP-CLI

Adds a `wp acm blueprint export` WP-CLI command to dump the current state of the site as a zipped ACM blueprint.

By default this saves:
- ACM models and fields.
- ACM taxonomies.
- ACM entries.
- WP core posts and pages.
- Taxonomy terms.
- Post tags.
- Post meta.
- Media linked to a post via an ACM media field or stored in `_thumbnail_id` post meta.
- ACM relationships.

https://wpengine.atlassian.net/browse/MTKA-1389

## Testing

Includes integration tests for export functionality.

### To test manually

1. Open a shell ready for WP-CLI. (In Local choose Open Site Shell from the site options menu.)
2. Run `wp acm blueprint export --open`.
3. You should see a directory containing your blueprint zip file. It will contain media if exported posts have post meta for a featured image or ACM media field.

To see all options use `wp help acm blueprint export`.

If you wish you can also try:

1. Taking the zip generated by the export.
2. Uploading it somewhere publicly accessible (such as to a WPE WP test site).
3. Removing existing ACM posts/taxonomies from your site. (`wp option delete atlas_content_modeler_post_types && wp option delete atlas_content_modeler_taxonomies`)
4. Restoring models/taxonomies/other state with `wp acm blueprint import https://example.com/path/to/your/blueprint.zip`.

## Screenshots

Defaults:

```sh
> wp acm blueprint export --open
Collecting ACM data.
Collecting ACM models.
Collecting ACM taxonomies.
Collecting posts.
Collecting terms.
Collecting post meta.
Collecting media.
Collecting ACM relationships
Writing acm.json manifest.
Generating zip.
Opening blueprint temp folder.
Success: Blueprint saved to /var/folders/mc/yf37tb4j5bjbwhv6z0m1bcfm0000gp/T/acm/acm-blueprint/acm-blueprint.zip.
```

With custom options:

```sh
> wp acm blueprint export --name="Custom Name" --description="My blueprint is great." --open
Collecting ACM data.
Collecting ACM models.
Collecting ACM taxonomies.
Collecting posts.
Collecting terms.
Collecting post meta.
Collecting media.
Collecting ACM relationships
Writing acm.json manifest.
Generating zip.
Opening blueprint temp folder.
Success: Blueprint saved to /var/folders/mc/yf37tb4j5bjbwhv6z0m1bcfm0000gp/T/acm/custom-name/custom-name.zip.
```

```sh
> wp help acm blueprint export

NAME

  wp acm blueprint export

DESCRIPTION

  Exports an ACM blueprint using the current state of the site.

SYNOPSIS

  wp acm blueprint export [--name] [--description] [--min-wp] [--min-acm] [--version] [--cpt-include] [--open]

  [--name]
    Optional blueprint name. Used in the manifest and zip file name.
    Defaults to “ACM Blueprint” resulting in acm-blueprint.zip.

  [--description]
    Optional description of the blueprint.

  [--min-wp]
    Minimum WordPress version. Defaults to current WordPress version.

  [--min-acm]
    Minimum Atlas Content Modeler plugin version. Defaults to current
    ACM version.

  [--version]
    Optional blueprint version. Defaults to 1.0.

  [--cpt-include]
    Post types to include, separated by commas. Defaults to post, page and
    all registered ACM post types.

  [--open]
    Open the folder containing the generated zip on success (macOS only,
    requires that `shell_exec()` has not been disabled).
```

## Documentation Changes

I plan to file a separate PR or add docs to Google Docs for internal use by other teams.

## Dependant PRs

n/a
